### DISCUSSION TITLE
BsdParser how to read *.bsd file and register to dataTypeDictionaryManager ?

### DISCUSSION THREAD

**Discussion by johannwesely** [eclipse-milo/milo#1168] [General]
Hi, i want to load a binary sheme Opc.Ua.Ijt.Base.Types.bsd from Companion Specification OPC UA for Joining Systems into our milo Server.
Now im not aware how to archive this. I have the Problem that i need the dataTypeId and encodingId which is not included in the bsd file.

`	public static void parseTypeDictionary(DataTypeDictionaryManager dataTypeDictionaryManager) throws Exception {
		BsdParser parser = new GenericBsdParser();

		DictionaryDescription dictionary = parser.parse(
				BSDParser.class.getClassLoader().getResourceAsStream(
						"Opc.Ua.Ijt.Base.Types.bsd")
		);

		dictionary.getStructCodecs().forEach(cd ->
						dataTypeDictionaryManager.getDictionary().registerStructCodec(cd.getCodec(), cd.getDescription(),??????,?????? )

		
		);
	}`  
I don't know where to get the dataTypeId and encodingId from i could read from .csv which comes with the companion spec but im not shure if this is the right way to handle this. Do you have any suggestions ?

**Comment by kevinherron** [comment]
You are on an unsupported path here, I'm not sure I can help you.

Even loading a companion spec into the server is not supported at this point (though it has been explored unofficially here: https://github.com/kevinherron/machine-example/tree/master).

I'm not sure what you're hoping to achieve with one of the `bsd` files. They are a deprecated mechanism for describing a type dictionary, and the only way they are used in Milo is by the client for discovering custom structures defined by a server. But defining and using custom types in the server via one of these files is probably not possible.

**Comment by johannwesely** [comment]
So this means when i encounter a struct in one of the specifications i have to define it like in the customStructType example. But there are alot of structs in the one i want to use right now. Also when i want to load another spec i have to define all included structs again. Which is huge amount of work. I think it should be somehow possible to load the structs from some sort of file if you plan in the future to support companion Specifications with the Server.

**Reply by kevinherron** [reply]
In the future companion specs will be supported via code generation, i.e. you will feed a UANodeSet XML model into a code generator and it will generate datatypes as well as object types and variable types for each of client and server, much like the built in datatypes, object types, and variable types that exist in Milo right now for the OPC UA namespace.

Support for this will likely come after version 1.0.0. The code generators exist but are WIP and not open source.

**Reply by johannwesely** [reply]
Thank you very much for your information.
